{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9cbb31da",
   "metadata": {},
   "source": [
    "# Build a model of a protein family starting from a single sequence\n",
    "\n",
    "\n",
    "1. Retrieve homologous (similar) sequences\n",
    "2. Build a MSA from retrieved sequences\n",
    "3. Caclulate a PSSM or HMM from the MSA\n",
    "4. Evaluate the model against a ground truth dataset\n",
    "\n",
    "\n",
    "### Implementation\n",
    "For each step some alternative implementations that can have an effect on the model accuracy\n",
    "\n",
    "**Step 1**\n",
    "- Blast against a database (SwissProt, UniProt, RefSeq, NR, ReferenceProteomes, ...)\n",
    "\n",
    "**Step 2**\n",
    "- Build the MSA from Blast alignments \n",
    "- Build MSA realigning Blast hits (ClustalO, T-Coffee, Muscle, ...)\n",
    "- Edit the MSA to remove poorly conserved columns\n",
    "- Remove sequence redundancy, similar sequences, to improve model generalization\n",
    "- Split the MSA, e.g. when there are multiple domains\n",
    "\n",
    "**Step 3**\n",
    "- Build a HMM using the HMMER command line \n",
    "- Build a PSSM using the Blast command line\n",
    "- Skip this step and use the MSA in step 4 with HMMSEARCH\n",
    "\n",
    "**Step 4**\n",
    "- Define a ground truth dataset (Pfam, CATH, keywords, manual curation, ...)\n",
    "- Calculate preicison, recall, MCC, balanced accuracy\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "339c26af",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Blast import NCBIXML\n",
    "import pandas as pd\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e55f32a7",
   "metadata": {},
   "source": [
    "# Generate a MSA\n",
    "\n",
    "Align P06621 with the [NCBI Blast webservice](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastp&PAGE_TYPE=BlastSearch&LINK_LOC=blasthome)\n",
    "\n",
    "Use to alternative databases\n",
    "-SwissProt\n",
    "-RefSeq selected\n",
    "\n",
    "Set \"Max target sequences\" to 5000\n",
    "When finish click on \"Download\" and select XML format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bf4b485c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_id</th>\n",
       "      <th>subject_id</th>\n",
       "      <th>query_len</th>\n",
       "      <th>query_seq</th>\n",
       "      <th>match_seq</th>\n",
       "      <th>subject_seq</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>identity</th>\n",
       "      <th>positive</th>\n",
       "      <th>gaps</th>\n",
       "      <th>eval</th>\n",
       "      <th>bit_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1545301071|ref|WP_126472623.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTAIAAVLATAF+AGTALAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTAIAAVLATAFMAGTALAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>410</td>\n",
       "      <td>414</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2129.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1543118946|ref|WP_126023768.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>MRPSIHRTAIAAVLATAFVAGTA-LAQKRDNVLFQAATDEQPAVIK...</td>\n",
       "      <td>MRPSIHRTA+AA+LA AFVA  A  AQKRDNVLFQAATDEQPAVIK...</td>\n",
       "      <td>MRPSIHRTALAALLAAAFVAPAATWAQKRDNVLFQAATDEQPAVIK...</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>416</td>\n",
       "      <td>407</td>\n",
       "      <td>411</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>2097.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1797811451|ref|WP_159279459.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTA+AAVLATAFVAG A AQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTAVAAVLATAFVAGGAWAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>393</td>\n",
       "      <td>408</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1992.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1480890214|ref|WP_119552493.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTA+AA+LATAF+A  A AQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>MRPSIHRTALAALLATAFLAPAAWAQKRDNVLFQAATDEQPAVIKT...</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>389</td>\n",
       "      <td>405</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1978.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1011471139|ref|WP_062367304.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>MRPSIHRT-AIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIK...</td>\n",
       "      <td>MRP   RT   A +           AQKRDNVLFQAATD QPAV+K...</td>\n",
       "      <td>MRPFNQRTMLAALLATALLAPAAGWAQKRDNVLFQAATDAQPAVLK...</td>\n",
       "      <td>1</td>\n",
       "      <td>415</td>\n",
       "      <td>1</td>\n",
       "      <td>416</td>\n",
       "      <td>374</td>\n",
       "      <td>393</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1938.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4996</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1510926788|ref|WP_122926114.1| ArgE/DapE fa...</td>\n",
       "      <td>415</td>\n",
       "      <td>EQPAVIKTLEKLVNIETGT----GDAEGIAAAGNFLEAELKNLGFT...</td>\n",
       "      <td>E+  +I  ++ L+ I++       D  G      +L+A+L  +G  ...</td>\n",
       "      <td>ERDELIGLVQDLIRIDSVNPYLDADGPGEREMAAYLQAKLVEMGLE...</td>\n",
       "      <td>39</td>\n",
       "      <td>261</td>\n",
       "      <td>6</td>\n",
       "      <td>223</td>\n",
       "      <td>70</td>\n",
       "      <td>117</td>\n",
       "      <td>19</td>\n",
       "      <td>2.135720e-12</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4997</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1460625799|ref|WP_116468663.1| hydrolase [S...</td>\n",
       "      <td>415</td>\n",
       "      <td>LVVGDNIVGKIKGRGGKNLLLMSHMDTVY-LKGILAKAPFRVEGDK...</td>\n",
       "      <td>L  G+++   ++      +L   HMDTV+    +     +R +G  ...</td>\n",
       "      <td>LAHGEHLHLVVRPEAPIQMLFTGHMDTVFGADHVFQHGQWRTDG-T...</td>\n",
       "      <td>89</td>\n",
       "      <td>411</td>\n",
       "      <td>78</td>\n",
       "      <td>396</td>\n",
       "      <td>93</td>\n",
       "      <td>150</td>\n",
       "      <td>18</td>\n",
       "      <td>2.144680e-12</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4998</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1718634206|ref|WP_145030192.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>LFQAATDEQPAVIKTLEKLVNIETGTGDAEGIAAAGNFLEAELKNL...</td>\n",
       "      <td>+ +A T E  A I  + +L++I   +G+ +G+A   +FL  +LK  ...</td>\n",
       "      <td>MARAKTSENKAAIDLVLELLSISGKSGEEKGVA---DFLVKKLKGA...</td>\n",
       "      <td>32</td>\n",
       "      <td>237</td>\n",
       "      <td>1</td>\n",
       "      <td>210</td>\n",
       "      <td>73</td>\n",
       "      <td>106</td>\n",
       "      <td>24</td>\n",
       "      <td>2.197150e-12</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4999</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1393435437|ref|WP_109798908.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>IAAVLATAFVAGTALAQKRDN----VLFQAATDEQPAVIKTLEKLV...</td>\n",
       "      <td>IA+ LA +  A  A+A+  ++     +      EQ   +K LE LV...</td>\n",
       "      <td>IASALALSCSAVQAMAKGPESGPEARMIATIDAEQTRTLKFLETLV...</td>\n",
       "      <td>10</td>\n",
       "      <td>339</td>\n",
       "      <td>7</td>\n",
       "      <td>361</td>\n",
       "      <td>96</td>\n",
       "      <td>151</td>\n",
       "      <td>27</td>\n",
       "      <td>2.322320e-12</td>\n",
       "      <td>190.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5000</th>\n",
       "      <td>sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...</td>\n",
       "      <td>gi|1188361627|ref|WP_085482951.1| M20/M25/M40 ...</td>\n",
       "      <td>415</td>\n",
       "      <td>AATDEQPAVIKTLEKLVNIETGTGD----AEGIAAAGNFLEAELKN...</td>\n",
       "      <td>AAT E   ++  L +LV I++   D      G AA  +F+   L  ...</td>\n",
       "      <td>AATGES-ELVSLLRRLVEIDSSNPDLGSTGRGEAAVADFITDWLAP...</td>\n",
       "      <td>35</td>\n",
       "      <td>413</td>\n",
       "      <td>13</td>\n",
       "      <td>394</td>\n",
       "      <td>104</td>\n",
       "      <td>180</td>\n",
       "      <td>37</td>\n",
       "      <td>2.334210e-12</td>\n",
       "      <td>189.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5001 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               query_id  \\\n",
       "0     sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "1     sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "2     sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "3     sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "4     sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "...                                                 ...   \n",
       "4996  sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "4997  sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "4998  sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "4999  sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "5000  sp|P06621|CBPG_PSES6 Carboxypeptidase G2 OS=Ps...   \n",
       "\n",
       "                                             subject_id  query_len  \\\n",
       "0     gi|1545301071|ref|WP_126472623.1| M20/M25/M40 ...        415   \n",
       "1     gi|1543118946|ref|WP_126023768.1| M20/M25/M40 ...        415   \n",
       "2     gi|1797811451|ref|WP_159279459.1| M20/M25/M40 ...        415   \n",
       "3     gi|1480890214|ref|WP_119552493.1| M20/M25/M40 ...        415   \n",
       "4     gi|1011471139|ref|WP_062367304.1| M20/M25/M40 ...        415   \n",
       "...                                                 ...        ...   \n",
       "4996  gi|1510926788|ref|WP_122926114.1| ArgE/DapE fa...        415   \n",
       "4997  gi|1460625799|ref|WP_116468663.1| hydrolase [S...        415   \n",
       "4998  gi|1718634206|ref|WP_145030192.1| M20/M25/M40 ...        415   \n",
       "4999  gi|1393435437|ref|WP_109798908.1| M20/M25/M40 ...        415   \n",
       "5000  gi|1188361627|ref|WP_085482951.1| M20/M25/M40 ...        415   \n",
       "\n",
       "                                              query_seq  \\\n",
       "0     MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...   \n",
       "1     MRPSIHRTAIAAVLATAFVAGTA-LAQKRDNVLFQAATDEQPAVIK...   \n",
       "2     MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...   \n",
       "3     MRPSIHRTAIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIKT...   \n",
       "4     MRPSIHRT-AIAAVLATAFVAGTALAQKRDNVLFQAATDEQPAVIK...   \n",
       "...                                                 ...   \n",
       "4996  EQPAVIKTLEKLVNIETGT----GDAEGIAAAGNFLEAELKNLGFT...   \n",
       "4997  LVVGDNIVGKIKGRGGKNLLLMSHMDTVY-LKGILAKAPFRVEGDK...   \n",
       "4998  LFQAATDEQPAVIKTLEKLVNIETGTGDAEGIAAAGNFLEAELKNL...   \n",
       "4999  IAAVLATAFVAGTALAQKRDN----VLFQAATDEQPAVIKTLEKLV...   \n",
       "5000  AATDEQPAVIKTLEKLVNIETGTGD----AEGIAAAGNFLEAELKN...   \n",
       "\n",
       "                                              match_seq  \\\n",
       "0     MRPSIHRTAIAAVLATAF+AGTALAQKRDNVLFQAATDEQPAVIKT...   \n",
       "1     MRPSIHRTA+AA+LA AFVA  A  AQKRDNVLFQAATDEQPAVIK...   \n",
       "2     MRPSIHRTA+AAVLATAFVAG A AQKRDNVLFQAATDEQPAVIKT...   \n",
       "3     MRPSIHRTA+AA+LATAF+A  A AQKRDNVLFQAATDEQPAVIKT...   \n",
       "4     MRP   RT   A +           AQKRDNVLFQAATD QPAV+K...   \n",
       "...                                                 ...   \n",
       "4996  E+  +I  ++ L+ I++       D  G      +L+A+L  +G  ...   \n",
       "4997  L  G+++   ++      +L   HMDTV+    +     +R +G  ...   \n",
       "4998  + +A T E  A I  + +L++I   +G+ +G+A   +FL  +LK  ...   \n",
       "4999  IA+ LA +  A  A+A+  ++     +      EQ   +K LE LV...   \n",
       "5000  AAT E   ++  L +LV I++   D      G AA  +F+   L  ...   \n",
       "\n",
       "                                            subject_seq  query_start  \\\n",
       "0     MRPSIHRTAIAAVLATAFMAGTALAQKRDNVLFQAATDEQPAVIKT...            1   \n",
       "1     MRPSIHRTALAALLAAAFVAPAATWAQKRDNVLFQAATDEQPAVIK...            1   \n",
       "2     MRPSIHRTAVAAVLATAFVAGGAWAQKRDNVLFQAATDEQPAVIKT...            1   \n",
       "3     MRPSIHRTALAALLATAFLAPAAWAQKRDNVLFQAATDEQPAVIKT...            1   \n",
       "4     MRPFNQRTMLAALLATALLAPAAGWAQKRDNVLFQAATDAQPAVLK...            1   \n",
       "...                                                 ...          ...   \n",
       "4996  ERDELIGLVQDLIRIDSVNPYLDADGPGEREMAAYLQAKLVEMGLE...           39   \n",
       "4997  LAHGEHLHLVVRPEAPIQMLFTGHMDTVFGADHVFQHGQWRTDG-T...           89   \n",
       "4998  MARAKTSENKAAIDLVLELLSISGKSGEEKGVA---DFLVKKLKGA...           32   \n",
       "4999  IASALALSCSAVQAMAKGPESGPEARMIATIDAEQTRTLKFLETLV...           10   \n",
       "5000  AATGES-ELVSLLRRLVEIDSSNPDLGSTGRGEAAVADFITDWLAP...           35   \n",
       "\n",
       "      query_end  subject_start  subject_end  identity  positive  gaps  \\\n",
       "0           415              1          415       410       414     0   \n",
       "1           415              1          416       407       411     1   \n",
       "2           415              1          415       393       408     0   \n",
       "3           415              1          415       389       405     0   \n",
       "4           415              1          416       374       393     1   \n",
       "...         ...            ...          ...       ...       ...   ...   \n",
       "4996        261              6          223        70       117    19   \n",
       "4997        411             78          396        93       150    18   \n",
       "4998        237              1          210        73       106    24   \n",
       "4999        339              7          361        96       151    27   \n",
       "5000        413             13          394       104       180    37   \n",
       "\n",
       "              eval  bit_score  \n",
       "0     0.000000e+00     2129.0  \n",
       "1     0.000000e+00     2097.0  \n",
       "2     0.000000e+00     1992.0  \n",
       "3     0.000000e+00     1978.0  \n",
       "4     0.000000e+00     1938.0  \n",
       "...            ...        ...  \n",
       "4996  2.135720e-12      190.0  \n",
       "4997  2.144680e-12      190.0  \n",
       "4998  2.197150e-12      190.0  \n",
       "4999  2.322320e-12      190.0  \n",
       "5000  2.334210e-12      189.0  \n",
       "\n",
       "[5001 rows x 15 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "# Parse the XML output\n",
    "data = []\n",
    "\n",
    "# WARNING: if the code below crashes, remove \"CREATE VIEW\" lines manually from the XML file\n",
    "\n",
    "blast_input = 'data/P06621_sprot.xml'  # SwissProt database\n",
    "blast_input = 'data/P06621_refseq_selected.xml'  # \"RefSeq selected\" database\n",
    "\n",
    "with open(blast_input) as f:\n",
    "    blast_records = NCBIXML.parse(f)\n",
    "\n",
    "    # Iterate Psiblast rounds\n",
    "    for blast_record in blast_records:\n",
    "        \n",
    "        # Iterate query alignments\n",
    "        query_id = blast_record.query\n",
    "        for i, alignment in enumerate(blast_record.alignments):\n",
    "            subject_id = alignment.title\n",
    "            \n",
    "            for hsp in alignment.hsps:\n",
    "                data.append((query_id,\n",
    "                                subject_id,\n",
    "                                blast_record.query_length,\n",
    "                                hsp.query,\n",
    "                                hsp.match,\n",
    "                                hsp.sbjct,\n",
    "                                hsp.query_start,\n",
    "                                hsp.query_end,\n",
    "                                hsp.sbjct_start,\n",
    "                                hsp.sbjct_end,\n",
    "                                hsp.identities,\n",
    "                                hsp.positives,\n",
    "                                hsp.gaps,\n",
    "                                hsp.expect,\n",
    "                                hsp.score))\n",
    "                \n",
    "df = pd.DataFrame(data, columns=[\"query_id\", \"subject_id\", \"query_len\",\n",
    "                              \"query_seq\", \"match_seq\", \"subject_seq\",\n",
    "                              \"query_start\", \"query_end\", \n",
    "                              \"subject_start\", \"subject_end\", \"identity\", \"positive\", \"gaps\", \"eval\", \"bit_score\"])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "62b254d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract the sequence of subject proteins (hits) aligned with the query to build the MSA\n",
    "\n",
    "input_path = Path(blast_input)\n",
    "msa = []\n",
    "\n",
    "with open(\"{}/{}_blast_msa.fasta\".format(input_path.parent, input_path.stem), \"w\") as fout:\n",
    "    for index, row in df.iterrows():\n",
    "        mapped_seq = [\"-\"] * blast_record.query_length  # Empty list of length = query_length\n",
    "        c = 0\n",
    "        if row[\"eval\"] < 0.001:\n",
    "        #     print(row)\n",
    "            for l_q, l_s in zip(row['query_seq'], row['subject_seq']):\n",
    "                if l_q != \" \" and l_q != '-':\n",
    "                    mapped_seq[row[\"query_start\"] + c -1] = l_s if l_s != \" \" else \"-\"\n",
    "                    c += 1\n",
    "            fout.write(\">{}\\n{}\\n\".format(row[\"subject_id\"], \"\".join(mapped_seq)))\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38be23f0",
   "metadata": {},
   "source": [
    "# Generate a better MSA\n",
    "\n",
    "Generate a new, possibly better, MSA starting from the full sequences of hits retrieved by Blast using the Clustal Omega method\n",
    "\n",
    "1. Extract the protein IDs from previous files\n",
    "\n",
    "```\n",
    "    # SwissProt IDs\n",
    "    awk -F\"|\" '/>/ {print $2}' data/P06621_sprot_blast_msa.fasta > tmp\n",
    "    \n",
    "    # RefSeq IDs (ex. WP_094475116.1)\n",
    "    awk -F\"|\" '/>/ {print $4}' data/P06621_refseq_selected_blast_msa.fasta > tmp2\n",
    "   \n",
    "``` \n",
    "\n",
    "2. Download the sequences from UniProt using the **Retrieve/ID mapping** web tool\n",
    "    * For SwissProt IDs select \"From --> UniProtKB, To UniProtKB. Download into **P06621_sprot_blast_mapped.fasta** \n",
    "    * For RefSeq selected \"From --> RefSeq Protein, To UniProtKB. Select \"View UniParc results\" to increase the number of retrieved sequences. Download into **P06621_refseq_selected_blast_mapped.fasta**\n",
    "\n",
    "\n",
    "3. Use [ClustalO](https://www.ebi.ac.uk/Tools/msa/clustalo/) to generate new MSAs. Select the *Fasta* output format and save into\n",
    "    * **P06621_sprot_blast_clustal.fasta**\n",
    "    * **P06621_refseq_selected_blast_clustal.fasta**\n",
    "\n",
    "\n",
    "4. Compare the Blast MSA and the ClustalO MSA in JalView\n",
    "\n",
    "```\n",
    "    java -jar jalview-all-2.11.1.4-j1.8.jar\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4dead30f",
   "metadata": {},
   "source": [
    "# Build PSSMs\n",
    "\n",
    "To build PSSM from a MSA file it is necesary to use the command line version of Blast\n",
    "\n",
    "#### Install Blast (~250 Mb)\n",
    "```\n",
    "wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.12.0+-x64-linux.tar.gz\n",
    "tar -xzf ncbi-blast-2.12.0+-x64-linux.tar.gz\n",
    "\n",
    "```\n",
    "\n",
    "#### Create a PSSM from a Fasta MSA \n",
    "\n",
    "Before running the command remove *gi|1774282670|ref|WP_153585536.1|* from the **data/P06621_refseq_selected_blast_msa.fasta** file, for some reasons it breaks the PSSM build\n",
    "\n",
    "The content of the file in the -subject option is irrelevant, just use a valid fasta file\n",
    "```\n",
    "ncbi-blast-2.12.0+/bin/psiblast -subject data/P06621.fasta -in_msa data/P06621_sprot_blast_msa.fasta -out_pssm data/P06621_sprot_blast_msa.pssm -out_ascii_pssm data/P06621_sprot_blast_msa.pssm_ascii\n",
    "\n",
    "ncbi-blast-2.12.0+/bin/psiblast -subject data/P06621.fasta -in_msa data/P06621_refseq_selected_blast_msa.fasta -out_pssm data/P06621_refseq_selected_blast_msa.pssm -out_ascii_pssm data/P06621_refseq_selected_blast_msa.pssm_ascii\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5ce28c7",
   "metadata": {},
   "source": [
    "# Build HMM\n",
    "\n",
    "Install HMMER (~18 Mb)\n",
    "```\n",
    "wget http://eddylab.org/software/hmmer/hmmer.tar.gz\n",
    "tar -xzf hmmer.tar.gz\n",
    "./configure\n",
    "make\n",
    "```\n",
    "\n",
    "Generate HMMs from MSAs\n",
    "```\n",
    "hmmer-3.3.2/src/hmmbuild data/P06621_sprot_blast_msa.hmm data/P06621_sprot_blast_msa.fasta\n",
    "\n",
    "hmmer-3.3.2/src/hmmbuild data/P06621_refseq_selected_blast_msa.hmm data/P06621_refseq_selected_blast_msa.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ba67f0",
   "metadata": {},
   "source": [
    "# Build HMM and PSSM from ClustalO MSAs\n",
    "\n",
    "Generate PSSMs\n",
    "```\n",
    "ncbi-blast-2.12.0+/bin/psiblast -subject data/P06621.fasta -in_msa data/P06621_sprot_blast_clustal.fasta -out_pssm data/P06621_sprot_blast_clustal.pssm -out_ascii_pssm data/P06621_sprot_blast_clustal.pssm_ascii\n",
    "\n",
    "ncbi-blast-2.12.0+/bin/psiblast -subject data/P06621.fasta -in_msa data/P06621_refseq_selected_blast_clustal.fasta -out_pssm data/P06621_refseq_selected_blast_clustal.pssm -out_ascii_pssm data/P06621_refseq_selected_blast_clustal.pssm_ascii\n",
    "```\n",
    "\n",
    "Generate HMMs\n",
    "```\n",
    "hmmer-3.3.2/src/hmmbuild data/P06621_sprot_blast_clustal.hmm data/P06621_sprot_blast_clustal.fasta\n",
    "\n",
    "hmmer-3.3.2/src/hmmbuild data/P06621_refseq_selected_blast_clustal.hmm data/P06621_refseq_selected_blast_clustal.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53aff5d2",
   "metadata": {},
   "source": [
    "# Evaluation\n",
    "\n",
    "Compare performance of your models in comparison with Pfam **Peptidase_M20.hmm**. \n",
    "\n",
    "1. Evaluate the number of hits retrieved by your models and those retrived by the Pfam domain aginst the SwissProt database.\n",
    "\n",
    "    * **PSSMs** You can use the web service. When you select Psi-Blast it is possible to provide a PSSM file as input\n",
    "\n",
    "    * **HMMs** You can use HMMSEARCH from HMMER web site\n",
    "\n",
    "\n",
    "2. Evaluate the overlap between the hits retrieved by your models and the hits retrived by the Pfam model\n",
    "\n",
    "    * You need to download sequence IDs of the matched sequences \n",
    "\n",
    "\n",
    "3. Evaluate False Positive matches (if any) and decide if they are good family candidates missed by Pfam but not by your model\n",
    "\n",
    "    * You need to look at alternative information to judge. Statistical parameters, curated annotations, etc...\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9f4347a",
   "metadata": {},
   "source": [
    "# Improve your model\n",
    "\n",
    "Manually edit the MSA in order to achieve better precision and/or recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dde1a49",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
